LibOS/shim/src/bookkeep/shim_vma.c

```diff
@@ -1347,66 +1078,39 @@ END_RS_FUNC(vma)
 
 BEGIN_CP_FUNC(all_vmas)
 {
-    struct shim_vma * tmp, ** vmas;
-    int nvmas = 0, cnt = 0;
-
-    lock(vma_list_lock);
-
-    __shrink_vmas();
-
-    listp_for_each_entry(tmp, &vma_list, list)
-        if (!(tmp->flags & VMA_INTERNAL))
-            nvmas++;
+    size_t count = DEFAULT_VMA_COUNT;
+    struct shim_vma_val * vmas = malloc(sizeof(*vmas) * count);
+    int ret;
 
-    if (!nvmas) {
-        unlock(vma_list_lock);
-        return 0;
-    }
+    if (!vmas)
+        return -ENOMEM;
 
-    vmas = __alloca(sizeof(struct shim_vam *) * nvmas);
+retry_dump_vmas:
+    ret = dump_all_vmas(vmas, count);
 
-    listp_for_each_entry(tmp, &vma_list, list)
-        if (!(tmp->flags & VMA_INTERNAL)) {
-            get_vma(tmp);
-            vmas[cnt++] = tmp;
+    if (ret == -EOVERFLOW) {
+        struct shim_vma_val * new_vmas
+                = malloc(sizeof(*new_vmas) * count * 2);
+        if (!new_vmas) {
+            free(vmas);
+            return -ENOMEM;
         }
-
-    unlock(vma_list_lock);
-
-    for (cnt = 0 ; cnt < nvmas ; cnt++) {
-        DO_CP(vma, vmas[cnt], NULL);
-        put_vma(vmas[cnt]);
+        free(vmas);
+        vmas = new_vmas;
+        count *= 2;
+        goto retry_dump_vmas;
     }
-}
-END_CP_FUNC_NO_RS(all_vmas)
 
-void debug_print_vma_list (void)
```

Comment:
I'd prefer to leave this function, it was very useful when debugging that race condition on the vma list.

Reply:
Fixed.